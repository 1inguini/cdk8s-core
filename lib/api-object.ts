import { Construct } from '@aws-cdk/core';
import { ObjectmetaMetaV1 } from './.generated/objectmeta-meta-v1';
import { removeEmpty } from './util';

export interface ApiObjectProps {
  /**
   * The API object spec.
   */
  readonly spec?: any;

  /**
   * Data associated with the resource.
   */
  readonly data?: any;

  /**
   * Resource kind.
   */
  readonly kind: string;

  /**
   * API version.
   */
  readonly apiVersion: string;

  /**
   * Object metadata.
   *
   * If `name` is not specified, an app-unique name will be allocated by the
   * framework based on the path of the construct within thes construct tree.
   */
  readonly metadata?: ObjectmetaMetaV1;
}

export class ApiObject extends Construct {

  /**
   * The app-unique name of the object.
   *
   * The name is allocated based on the path of the object construct within the
   * construct tree.
   */
  public readonly name: string;

  public readonly apiVersion: string;
  public readonly kind: string;

  private readonly metadata: ObjectmetaMetaV1;

  constructor(scope: Construct, id: string, private readonly props: ApiObjectProps) {
    super(scope, id);

    this.kind = props.kind;
    this.apiVersion = props.apiVersion;

    this.metadata = props.metadata || {};
    this.metadata.name = this.metadata.name || this.node.uniqueId.toLowerCase();
    this.name = this.metadata.name;
  }

  /**
   * Renders the object to Kubernetes config.
   */
  public render(): any {
    return removeEmpty({
      apiVersion: this.props.apiVersion,
      kind: this.props.kind,
      metadata: this.metadata,
      data: this.props.data,
      spec: this.props.spec
    });
  }
}

