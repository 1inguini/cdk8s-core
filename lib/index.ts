export * from './service';
export * from './deployment';
export * from './resource';
export * from './stack';
