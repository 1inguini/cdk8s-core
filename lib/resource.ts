import { Construct } from '@aws-cdk/core';
import { ObjectmetaMetaV1 } from './.generated/objectmeta-meta-v1';

export interface KubResourceProps {
  /**
   * The resource API spec.
   */
  readonly spec?: any;

  /**
   * Data associated with the resource.
   */
  readonly data?: any;

  /**
   * Resource kind.
   */
  readonly kind: string;

  /**
   * API version.
   */
  readonly apiVersion: string;

  /**
   s* Object metadata.
   *
   * If `name` is not specified, an app-unique name will be allocated by the
   * framework based on the path of the construct within thes construct tree.
   */
  readonly metadata?: ObjectmetaMetaV1;
}

export class KubResource extends Construct {

  /**
   * The app-unique name of the resource.
   *
   * The name is allocated based on the path of the resource construct within the
   * construct tree.
   */
  public readonly name: string;

  private readonly metadata: ObjectmetaMetaV1;

  constructor(scope: Construct, id: string, private readonly props: KubResourceProps) {
    super(scope, id);

    this.metadata = props.metadata || {};
    this.metadata.name = this.metadata.name || this.node.uniqueId.toLowerCase();
  }

  /**
   * Renders the resource to Kubernetes config.
   */
  public render(): any {
    return removeEmpty({
      apiVersion: this.props.apiVersion,
      kind: this.props.kind,
      metadata: this.metadata,
      data: this.props.data,
      spec: this.props.spec
    });
  }
}

function removeEmpty(obj: any) {
  Object.entries(obj).forEach(([ key, value ]) => {
    if (!Array.isArray(obj) && typeof value === 'object') {
      removeEmpty(value);
    }
    else if (value == null) {
      delete obj[key];
    }
  });
  return obj;
};