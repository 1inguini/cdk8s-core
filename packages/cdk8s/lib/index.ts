export * from './service';
export * from './deployment';
export * from './api-object';
export * from './chart';
export * from './horizontal-pod-autoscaler';
export * from './ingress';