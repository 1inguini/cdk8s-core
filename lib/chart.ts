import { Construct, ISynthesisSession } from '@aws-cdk/core';
import fs = require('fs');
import path = require('path');
import { ApiObject } from './api-object';
const yaml = require('js-yaml');


export class Chart extends Construct {

  /**
   * The name of the stack's YAML file as emitted into the cloud assembly
   * directory during synthesis.
   */
  public readonly assemblyFileName: string;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.assemblyFileName = `${this.node.uniqueId}.k8s.yaml`;
  }

  protected synthesize(session: ISynthesisSession) {
    const resources = new Array<any>();

    for (const resource of this.node.findAll().filter(x => x instanceof ApiObject)) {
      resources.push((resource as ApiObject).render());
    }

    const doc = resources.map(r => yaml.dump(r, { skipInvalid: true })).join('---\n');
    fs.writeFileSync(path.join(session.assembly.outdir, this.assemblyFileName), doc);
  }
}
