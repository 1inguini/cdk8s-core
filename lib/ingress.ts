import { ApiObject } from './api-object';
import { Construct } from '@aws-cdk/core';
import { IngressspecExtensionsV1Beta1 } from './.generated/ingressspec-extensions-v1beta1';
import { ObjectmetaMetaV1 } from './.generated/objectmeta-meta-v1';

export interface IngressObjectOptions {
  readonly spec: IngressspecExtensionsV1Beta1;
  readonly metadata: ObjectmetaMetaV1;
}

export class IngressObject extends ApiObject {
  constructor(scope: Construct, id: string, options: IngressObjectOptions) {
    super(scope, id, {
      apiVersion: 'extensions/v1beta1',
      kind: 'Ingress',
      spec: options.spec,
      metadata: options.metadata
    });
  }
}