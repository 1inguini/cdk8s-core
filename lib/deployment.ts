import { KubResource } from './resource';
import { Construct } from '@aws-cdk/core';
import { DeploymentspecAppsV1 } from './.generated/deploymentspec-apps-v1';

export class KubDeployment extends KubResource {
  constructor(scope: Construct, ns: string, spec: DeploymentspecAppsV1) {
    super(scope, ns, {
      apiVersion: 'apps/v1',
      kind: 'Deployment',
      spec: spec
    });
  }
}
