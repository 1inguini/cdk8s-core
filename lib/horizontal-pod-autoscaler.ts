import { KubResource } from './resource';
import { Construct } from '@aws-cdk/core';
import { HorizontalpodautoscalerAutoscalingV2Beta1 } from './.generated/horizontalpodautoscaler-autoscaling-v2beta1';
import { ObjectmetaMetaV1 } from './.generated/objectmeta-meta-v1';

export interface KubHorizontalPodAutoscalerOptions {
  readonly spec: HorizontalpodautoscalerAutoscalingV2Beta1;
  readonly metadata?: ObjectmetaMetaV1;
}

export class KubHorizontalPodAutoscaler extends KubResource {
  constructor(scope: Construct, ns: string, options: KubHorizontalPodAutoscalerOptions) {
    super(scope, ns, {
      apiVersion: 'apps/v1',
      kind: 'Deployment',
      spec: options.spec,
      metadata: options.metadata
    });
  }
}
