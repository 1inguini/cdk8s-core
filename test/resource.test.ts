import { KubResource, Stack } from '../lib';
import { App, Construct } from '@aws-cdk/core';
import { synth } from './util';

test('minimal configuration', () => {
  const app = new App();
  const stack = new Stack(app, 'test');

  new KubResource(stack, 'MyResource', {
    apiVersion: 'v1',
    kind: 'MyResource'
  });

  expect(synth(stack)).toMatchSnapshot();
});

test('synthesized resource name is based on path', () => {
  // GIVEN
  const app = new App();
  const stack = new Stack(app, 'test');
  new KubResource(stack, 'MyResource', {
    apiVersion: 'v1',
    kind: 'MyResource'
  });

  // WHEN
  // now create another resource with the same namespace but within a sub-scope
  const scope = new Construct(stack, 'Scope');
  new KubResource(scope, 'MyResource', {
    apiVersion: 'v1',
    kind: 'MyResource'
  });

  // THEN
  expect(synth(stack)).toMatchSnapshot();
});

test('if name is explicitly specified it will be respected', () => {
  // GIVEN
  const app = new App();
  const stack = new Stack(app, 'test');

  // WHEN
  new KubResource(stack, 'MyResource', {
    kind: 'MyResource',
    apiVersion: 'v1',
    metadata: {
      name: 'boom'
    }
  });

  // THEN
  expect(synth(stack)).toMatchSnapshot();
});

test('"spec" is synthesized as-is', () => {
  // GIVEN
  const app = new App();
  const stack = new Stack(app, 'test');

  // WHEN
  new KubResource(stack, 'resource', {
    kind: 'ResourceType',
    apiVersion: 'v1',
    spec: {
      prop1: 'hello',
      prop2: {
        world: 123
      }
    }
  });

  // THEN
  expect(synth(stack)).toMatchSnapshot();
});

test('"data" can be used to specify resource data', () => {
  // GIVEN
  const app = new App();
  const stack = new Stack(app, 'test');

  // WHEN
  new KubResource(stack, 'resource', {
    kind: 'ResourceType',
    apiVersion: 'v1',
    data: {
      boom: 123
    }
  });

  // THEN
  expect(synth(stack)).toMatchSnapshot();
});