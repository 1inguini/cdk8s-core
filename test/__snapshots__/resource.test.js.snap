// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"data" can be used to specify resource data 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "data": Object {
      "boom": 123,
    },
    "kind": "ResourceType",
    "metadata": Object {
      "name": "testresource368fc2ec",
    },
  },
]
`;

exports[`"spec" is synthesized as-is 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "ResourceType",
    "metadata": Object {
      "name": "testresource368fc2ec",
    },
    "spec": Object {
      "prop1": "hello",
      "prop2": Object {
        "world": 123,
      },
    },
  },
]
`;

exports[`if name is explicitly specified it will be respected 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "MyResource",
    "metadata": Object {
      "name": "boom",
    },
  },
]
`;

exports[`minimal configuration 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "MyResource",
    "metadata": Object {
      "name": "testmyresource33fc126f",
    },
  },
]
`;

exports[`synthesized resource name is based on path 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "MyResource",
    "metadata": Object {
      "name": "testmyresource33fc126f",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "MyResource",
    "metadata": Object {
      "name": "testscopemyresourcec364662d",
    },
  },
]
`;
